### PR TITLE
Added Docker Support to Talawa Admin, Fixes #1075

### LINKED ISSUE #1075: Talawa-Admin Docker Support Issue
### Talawa-Admin Docker Support Issue:

**Title:**
Adding Docker Support to Talawa-Admin

**Description:**

**Problem:**
Talawa-Admin currently lacks Docker support, hindering its ease of deployment and scalability. Adding Docker support is essential to streamline the deployment process and make it more accessible to a broader audience.

**Solution:**
Implementing Docker support for Talawa-Admin would involve creating a Dockerfile and configuration to package the application into a Docker container. This containerization would simplify deployment, improve consistency across environments, and enhance scalability.

**Approach:**
To achieve Docker support for Talawa-Admin, we can follow these steps:
1. Create a Dockerfile specifying the necessary dependencies and configuration.
2. Configure the application to work seamlessly within a Docker container.
3. Document the steps for building and running Talawa-Admin using Docker in the project's README.

**Additional Context:**
Adding Docker support aligns with modern development practices, making it easier for developers to set up and run Talawa-Admin in various environments. Docker containers provide isolation, portability, and reproducibility, contributing to a more robust and user-friendly development and deployment experience.

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by vasujain275**
**What kind of change does this PR introduce?**

This PR add Docker Support to Talawa Admin.

**Issue Number:**

- Closes #1075 

**Did you add tests for your changes?**

N/A

**Snapshots/Videos:**

**If relevant, did you update the documentation?**

N/A

**Summary**

**1. Introduction of Docker Support:**
- This PR adds Docker support to Talawa Admin.
- This enhancement aims to facilitate a smoother developer experience, especially for individuals contributing to Talawa API and Talawa App.

**2. Consideration of Docker Methods:**
- Initially, the idea was to implement two Docker methods - dev and production.
- However, upon scrutinizing the codebase, it was discovered that we are utilizing `Create React App` and react scripts for the build process.
- The current build process is evidently not optimized for either production-ready minified binaries or Docker hot reloading.

**3. Proposal for Build Process Improvement:**
- To address this limitation, I think we should migrate to a more suitable build process like Vite.
- This migration is necessary to enable efficient building of both minified production-ready build files and Docker hot reloading.

**4. Planned Actions:**
- After this PR is merged, I will raise an issue to migrate from React scripts to the Vite build process, we can have further discussion on build process there.
- This action will lay the groundwork for future improvements in better and improved Docker support once the Build Process is sorted.

**5. Current Status:**
- Currently, separate dev and prod Docker methods have not been implemented due to the inadequacy of the existing build process.
- The current build process cannot produce minified production-ready build files, and a dev hot reloading Docker setup would be impractical due to the same issue.

**Does this PR introduce a breaking change?**

No

**Other information**

You can test the current Docker Installtion process by using - 
```bash
npm run dx-prod
```

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

- **New Features**
  - Introduced a Docker setup for the application.
  - Added a new script for Docker operations (`dx-prod`).

- **Chores**
  - Updated `.dockerignore` and `.eslintignore` files to optimize builds and linting.
  - Updated `package.json` with new and updated dependencies (`dotenv`, `inquirer`).

- **Configuration**
  - Updated `docker-compose.yml` to configure the application service, set environment variables, and expose port 4321.

### CODE REVIEW

**Comment by palisadoes** [comment]
## The PR merging freeze is lifted.

1. We are working on bug fixes that may arise. These should be few and are being tracked here:
    1. https://github.com/orgs/PalisadoesFoundation/projects/24
1. Please update your PRs and local repos. Fix any new merge conflicts that may have occurred.
    
Background:

1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by palisadoes** [comment]
## The PR merging freeze is lifted.

1. We are working on bug fixes that may arise. These should be few and are being tracked here:
    1. https://github.com/orgs/PalisadoesFoundation/projects/24
1. Please update your PRs and local repos. Fix any new merge conflicts that may have occurred.
    
Background:

1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by Cioppolo14** [comment]
@vasujain275 Can you fix the linting test issue?

**Comment by vasujain275** [comment]
@Cioppolo14 Sorry for the late response, I was not well. Can you link me to the issue you are referring to?

**Comment by palisadoes** [comment]
1. There is a linting test that is failing. You need to submit your PR using a different source branch. This requirement was implemented 2-3 weeks ago
2. Either close this PR or update it with a source branch with a different name.

**Comment by Cioppolo14** [comment]
@vasujain275 Please fix the linting issue.

**Comment by vasujain275** [comment]
> 1. There is a linting test that is failing. You need to submit your PR using a different source branch. This requirement was implemented 2-3 weeks ago
> 2. Either close this PR or update it with a source branch with a different name.

I tried solving the liniting issue but not able to understand how to solve it, any guidance on how to do it would be really helpful. Otherwise I can close this PR and create a new one if solving the linting issue in this pr is too complex.

**Comment by palisadoes** [comment]
Is anything else required for a review of this?

**Comment by vasujain275** [comment]
> Is anything else required for a review of this?

No, I don't think so

**Review by varshith257 — COMMENTED** [review]
@vasujain275 Add ` dockerignore`  file

**Comment by vasujain275** [comment]
> @vasujain275 Add ` dockerignore` file
> 
> cc: @Cioppolo14

Okay

**Comment by vasujain275** [comment]
@varshith257 Added .dockerignore file, thanks for the suggestion

**Comment by palisadoes** [comment]
@vasujain275 The tests aren't running. PTAL

**Comment by vasujain275** [comment]
Okay, let me take a look

**Comment by vasujain275** [comment]
Fixed the conflicting files

**Comment by palisadoes** [comment]
@varshith257 PTAL here too. We need to get this functionality working so that we have a live demo instance of the API and Admin in the next few weeks. This will help with us getting grants for internships as well as new contributors

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+FROM node:20.10.0 AS build
+
+WORKDIR /usr/src/app
+
+COPY . .
+
+RUN npm install -g typescript
+
+RUN npm install
```
[the remark this answers is not in the document]
Comment (vasujain275) [reply Dockerfile:9]: They are not supposed to be run together in my opinion

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+RUN npm run build
```
[the remark this answers is not in the document]
Comment (vasujain275) [reply Dockerfile:11]: These both are very different commands and do not affect docker image performance

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,15 @@
+FROM node:20.10.0 AS build
+
+WORKDIR /usr/src/app
+
+COPY . .
+
+RUN npm install -g typescript
```
[the remark this answers is not in the document]
Comment (vasujain275) [reply Dockerfile:7]: Installing typescript globally is better I think for stability reasons and anyway doesn't affect the image performance

**Review by varshith257 — COMMENTED** [review]
What's the current image size of it?

File: package.json [hunk 4]
```diff
@@ -31,7 +32,7 @@
     "i18next": "^21.8.14",
     "i18next-browser-languagedetector": "^6.1.4",
     "i18next-http-backend": "^1.4.1",
-    "inquirer": "^8.0.0",
```
[the remark this answers is not in the document]
Comment (vasujain275) [reply package.json:34]: This package higher version was needed for the build process to work.

**Comment by palisadoes** [comment]
Please fix the failing workflow

**Comment by palisadoes** [comment]
Should this PR be closed? It's been open a long time.

**Comment by vasujain275** [comment]
> Should this PR be closed? It's been open a long time.

Forgot about this PR entirely, my bad. I glanced through it and it is a much needed change to our docker files of projects. 

Let me fix the conflicts and finalize this for review by this or at max next weekend.

**Comment by noman2002** [comment]
Closing due to inactivity.

**Pull request closed without merging**